src/workerd/api/eventsource.c++

```diff
@@ -0,0 +1,520 @@
+    // cr            = %x000D ; U+000D CARRIAGE RETURN (CR)
+    // space         = %x0020 ; U+0020 SPACE
+    // colon         = %x003A ; U+003A COLON (:)
+    // bom           = %xFEFF ; U+FEFF BYTE ORDER MARK
+    // name-char     = %x0000-0009 / %x000B-000C / %x000E-0039 / %x003B-10FFFF
+    //                 ; a scalar value other than U+000A LINE FEED (LF), U+000D CARRIAGE RETURN
+    //                   (CR), or U+003A COLON (:)
+    // any-char      = %x0000-0009 / %x000B-000C / %x000E-10FFFF
+    //                 ; a scalar value other than U+000A LINE FEED (LF) or U+000D CARRIAGE
+    //                   RETURN (CR)
+
+    // Note that the BOM (if present) is filtered out in the write() method.
+
+    if (line.size() == 0) {
+      // Dispatch the current pending message and clear it. If there is no
+      // pending message, we'll just ignore the line.
+      KJ_IF_SOME(pending, currentPendingMessage) {
+        // This message is done and ready to be dispatched. Add it to the
+        // pendingMessages list. The next time release() is called, it will
+        // be passed off to the EventSource.
+        pending.id = kj::str(KJ_ASSERT_NONNULL(eventSource).getLastEventId());
+        pendingMessages.add(kj::mv(pending));
+        currentPendingMessage = kj::none;
+      }
+    } else if (line[0] == ':') {
+      // Ignore the line.
+    } else {
+      static constexpr auto handle =
+          [](auto& self, kj::ArrayPtr<const char> field, kj::ArrayPtr<const char> value) {
+        auto& pending = self.getPendingMessage();
+        auto& ev = KJ_ASSERT_NONNULL(self.eventSource);
+        // Per the spec, only one space after the colon is optional and trimmed.
+        // Any other whitespace, or additional spaces aren't accounted for so would
+        // be part of the value.
+        if (value.size() > 0 && value[0] == ' ') {
+          value = value.slice(1);
+        }
+        if (field == "data"_kjc) {
+          pending.data.add(kj::str(value));
+        } else if (field == "event"_kjc) {
+          pending.event = kj::str(value);
+        } else if (field == "id"_kjc) {
+          ev.setLastEventId(kj::str(value));
+        } else if (field == "retry"_kjc) {
+          KJ_IF_SOME(time, kj::str(value).tryParseAs<uint32_t>()) {
+            KJ_ASSERT_NONNULL(self.eventSource).setReconnectionTime(time);
+          }
+          // Ignore the line if it cannot be successfully parsed as a uint32_t
+        }
+      };
+
+      KJ_IF_SOME(pos, line.findFirst(':')) {
+        handle(*this, line.slice(0, pos), line.slice(pos + 1));
+      } else {
+        handle(*this, line, ""_kjc);
+      }
+    }
+  }
+
+  void release() {
+    if (pendingMessages.size() == 0) return;
+    auto pending = pendingMessages.releaseAsArray();
+    // If the event source is gone, just drop the messages on the floor.
+    KJ_IF_SOME(es, eventSource) {
+      es.enqueueMessages(kj::mv(pending));
+    }
+  }
+
+  void clear() {
```

Comment:
Do we need to do anything for the `pendingMessage`?

Reply:
Not sure what you mean. We are clearing the `pendingMessages`

Reply:
Sorry, I meant the `currentPendingMessage`.

Reply:
ah right. added